src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java

```diff
@@ -0,0 +1,312 @@
+		    // assume WN profile: single PEM encoded chain. Should also
+		    // be single-valued according to https://edms.cern.ch/document/1058175/1.0.1
+		    if (attributeValues.size() > 1)
+			log.warn("key-info attr has >1 values, but is PEM, taking only first value.");
+		} else {
+		    // assume CommonXACMLProfile: base64 encoded blobs, might
+		    // not be ordered, according to
+		    // https://twiki.cern.ch/twiki/bin/view/EMI/CommonXACMLProfileV1_1
+		    // so combine into PEM string ourselves and let canl-java
+		    // loadCertificateChain() sort it for us.
+		    StringBuilder pem = new StringBuilder();
+		    for (Object attrval: attributeValues)    {
+			pem.append("-----BEGIN CERTIFICATE-----\n");
+			pem.append((String)attrval);
+			pem.append("\n-----END CERTIFICATE-----\n");
+		    }
+		    value = pem.toString();
+		}
+		// Now convert PEM-string value to a X509Certificate
+		InputStream pemReader = new ByteArrayInputStream(value.getBytes(StandardCharsets.UTF_8));
+		try {
+		    // pemReader is closed by loadCertificateChain()
+		    X509Certificate[] chain = CertificateUtils.loadCertificateChain(pemReader, Encoding.PEM);
+		    return ProxyUtils.getEndUserCertificate(chain);
+		} catch (Exception e) {
+		    // This might be a IOException, but also for invalid base64
+		    // a StringIndexOutOfBoundsException or a DecoderException
+		    log.error("Skipping invalid key-info attr: Parsing value as a certificate failed: "+e.getMessage());
+		}
+	    }
+	}
+	// No cert found
+	log.info("No valid certificate found in set of attributes");
+	return null;
+    }
+
+    /**
+     * Tries to obtain policy OIDs from end-entity certificate
+     * @param cert input (end-entity) certificate
+     * @return String array of policy OIDs
+     */
+    private String[] getCAPolicyOids(X509Certificate cert)  {
+	List<String> oidList = new LazyList<String>();
+
+	// OID for certificate_policies (=2.5.29.32)
+//	String certPolicies = X509Extension.certificatePolicies.toString();
+	String certPolicies = Extension.certificatePolicies.toString();
+
+	// Grab bare extension value from certificate
+	byte[] extvalue = cert.getExtensionValue(certPolicies);
+	if (extvalue==null)
+	    return null;
+
+	// Try to parse the raw bytes as ASN1Sequence
+	ASN1Sequence seq;
+	try {
+	    DEROctetString oct=(DEROctetString)(new ASN1InputStream(new ByteArrayInputStream(extvalue)).readObject());
+	    seq = (ASN1Sequence) new ASN1InputStream(new ByteArrayInputStream(oct.getOctets())).readObject();
+	} catch (IOException e)	{
+	    log.error("Trying to obtain policyinfo from certificate failed: "+e.getMessage());
+	    return null;
+	}
+
+	// Parse the ASN1Sequence as policy information values
+	for (int pos = 0; pos < seq.size(); pos++) {
+	    PolicyInformation policyInfo = PolicyInformation.getInstance(seq.getObjectAt(pos));
+	    oidList.add(policyInfo.getPolicyIdentifier().getId());
+	}
+
+	// Return oidList
+	return oidList.toArray(new String[0]);
```

Comment:
Not any more apparently: http://stackoverflow.com/questions/4042434/converting-arrayliststring-to-string-in-java/4042464#4042464 see also the bottomline of https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_conclusion